File: 백준/Silver/1325. 효율적인 해킹/효율적인 해킹.js
const fs = require('fs');
const stdin = fs.readFileSync('/dev/stdin').toString().trim();

// -------------
// 입력
// -------------
const input = stdin.split('\n').map(item => item.trim().split(' ').map(Number));
const [[N, M], ...CONN] = input;

// -------------
// 풀이 (BFS)
// -------------
function solution1(n, m, conn) {
  const graph = Array.from({ length: n + 1 }, () => []);
  for (let [a, b] of conn) {
    graph[b].push(a);
  }
  const visited = Array(n + 1);
  
  const bfs = (com) => {
    const stack = [com];
    visited[com] = 1;
    let cnt = 1;

    while (stack.length) {
      const now = stack.pop();
      for (let i = 0; i < graph[now].length; i++) {
        if (!visited[graph[now][i]]) {
          stack.push(graph[now][i]);
          visited[graph[now][i]] = 1;
          cnt++;
        }
      }
    }
    if (maxCnt < cnt) {
      maxCnt = cnt;
      result = [com];
    } else if (maxCnt === cnt) {
      result.push(com);
    }
  }

  let maxCnt = 0;
  let result = [];
  for (let i = 1; i <= n; i++) {
    visited.fill(0);
    bfs(i);
  }
  return result.join(' ');
}

// -------------
// 출력
// -------------
const result = solution1(N, M, CONN);
console.log(result);

File: 백준/Silver/1325. 효율적인 해킹/효율적인 해킹.py
from collections import deque
import sys
read = sys.stdin.readline

# -------------
#  입력
# -------------
N, M = [*map(int, read().rstrip().split())]
COMPUTERS = [[*map(int, read().rstrip().split())] for _ in range(M)]

# -------------
#  풀이 BFS
# -------------
graph = [[] for _ in range(N+1)]
for a, b in COMPUTERS:
  graph[b].append(a)

def bfs(startCom):
  queue = deque([startCom])
  visited = [0] * (N+1)
  visited[startCom] = 1

  cnt = 0
  while queue:
    com = queue.popleft()
    
    cnt += 1
    for next in graph[com]:
      if visited[next]: continue
      visited[next] = 1
      queue.append(next)
  
  return cnt

answer = []
maxCnt = 0
for com in range(1, N+1):
  cnt = bfs(com)
  if maxCnt == cnt:
    answer.append(com)
    
  if maxCnt < cnt:
    maxCnt = cnt
    answer = [com]

print(*sorted(answer))